File: plugins/wise/engine/src/steps/bash.ts
// `bash` step: `bash -c <run>` in the run cwd under the P6 clean env, wall-clock timeout,
// whole trimmed stdout as the single declared output, last stdout line as the verdict.

import { cleanEnv, spawnClean } from "../adapters/spawn.ts";
import type { Env } from "../paths.ts";
import type { BashStep } from "../types.ts";
import { DEFAULT_STEP_TIMEOUT_MS, headline } from "./agent.ts";

const STDERR_TAIL = 800;
const STDOUT_CAP = 1024 * 1024;

export type BashResult = {
  ok: boolean;
  code: number | null;
  timedOut: boolean;
  stdout: string;
  stderr: string;
  verdict: string;
  error?: string;
  outputs: Record<string, string>;
};

export type BashHandle = {
  pid: number;
  kill: (signal?: NodeJS.Signals) => void;
  result: Promise<BashResult>;
};

export type BashStepOpts = {
  cwd: string;
  parentEnv?: Env;
  defaultTimeoutMs?: number;
};

function lastLine(text: string): string {
  const lines = text
    .split("\n")
    .map((l) => l.trim())
    .filter((l) => l.length > 0);
  return lines.at(-1) ?? "";
}

/** Start the step; `kill` is available at once, `result` settles on exit. */
export function startBashStep(step: BashStep, opts: BashStepOpts): BashHandle {
  const timeoutMs =
    step.timeout !== undefined
      ? step.timeout * 1000
      : (opts.defaultTimeoutMs ?? DEFAULT_STEP_TIMEOUT_MS);
  const env = cleanEnv(opts.parentEnv ? { parent: opts.parentEnv } : {});
  const proc = spawnClean("bash", ["-c", step.run], { cwd: opts.cwd, env, timeoutMs });
  let stdout = "";
  proc.stdout.on("data", (chunk: string) => {
    if (stdout.length < STDOUT_CAP) stdout += chunk.slice(0, STDOUT_CAP - stdout.length);
  });
  proc.stdin.end();
  const result = proc.exited.then((exit): BashResult => {
    const trimmed = stdout.trim();
    const ok = exit.code === 0 && !exit.timedOut && exit.error === undefined;
    const outputs: Record<string, string> = {};
    const name = step.outputs?.[0];
    if (ok && name !== undefined) outputs[name] = trimmed;
    if (ok) {
      return {
        ok,
        code: exit.code,
        timedOut: false,
        stdout: trimmed,
        stderr: exit.stderr,
        verdict: headline(lastLine(trimmed)) || "ok",
        outputs,
      };
    }
    const tail = exit.stderr.trim().slice(-STDERR_TAIL);
    const error = exit.timedOut
      ? `timed out after ${timeoutMs} ms`
      : exit.error !== undefined
        ? `spawn failed: ${exit.error}`
        : tail || `exit code ${String(exit.code)}${exit.signal ? ` (${exit.signal})` : ""}`;
    return {
      ok,
      code: exit.code,
      timedOut: exit.timedOut,
      stdout: trimmed,
      stderr: exit.stderr,
      verdict: headline(`failed: ${lastLine(error) || error}`),
      error,
      outputs,
    };
  });
  return { pid: proc.pid, kill: proc.kill, result };
}

export function runBashStep(step: BashStep, opts: BashStepOpts): Promise<BashResult> {
  return startBashStep(step, opts).result;
}

File: plugins/wise/engine/wise_engine/steps/bash.py
from __future__ import annotations

import asyncio
import codecs
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from ..scheduler import JS_WHITESPACE
from ..spawn import SpawnOptions, clean_env, spawn_clean
from .agent import DEFAULT_STEP_TIMEOUT_MS, headline, utf16_length, utf16_slice

STDERR_TAIL = 800
STDOUT_CAP = 1024 * 1024
Json = dict[str, Any]


@dataclass
class BashHandle:
    pid: int
    kill: Callable[..., Any]
    result: asyncio.Task[Json]


def _last_line(text: str) -> str:
    return next(
        (
            line.strip(JS_WHITESPACE)
            for line in reversed(text.split("\n"))
            if line.strip(JS_WHITESPACE)
        ),
        "",
    )


async def start_bash_step(step: Json, opts: Json) -> BashHandle:
    timeout_ms = (
        step["timeout"] * 1000
        if "timeout" in step
        else opts.get("default_timeout_ms", DEFAULT_STEP_TIMEOUT_MS)
    )
    proc = await spawn_clean(
        "bash",
        ["-c", step["run"]],
        SpawnOptions(
            cwd=opts["cwd"], env=clean_env(parent=opts.get("parent_env")), timeout_ms=timeout_ms
        ),
    )
    proc.stdin.end()

    async def finish() -> Json:
        chunks = []
        remaining = STDOUT_CAP
        decoder = codecs.getincrementaldecoder("utf-8")("replace")
        try:
            while chunk := await proc.stdout.read(65536):
                text = decoder.decode(chunk)
                if remaining:
                    clipped = utf16_slice(text, 0, remaining)
                    chunks.append(clipped)
                    remaining -= utf16_length(clipped)
            final = decoder.decode(b"", final=True)
            if remaining:
                chunks.append(utf16_slice(final, 0, remaining))
            exit = await proc.exited
        except asyncio.CancelledError:
            proc.exited.cancel()
            await asyncio.gather(proc.exited, return_exceptions=True)
            raise
        stdout = "".join(chunks).strip(JS_WHITESPACE)
        ok = exit.code == 0 and not exit.timed_out and exit.error is None
        outputs = {}
        if ok and step.get("outputs"):
            outputs[step["outputs"][0]] = stdout
        result = dict(
            ok=ok,
            code=exit.code,
            timed_out=exit.timed_out,
            stdout=stdout,
            stderr=exit.stderr,
            outputs=outputs,
        )
        if ok:
            return {**result, "verdict": headline(_last_line(stdout)) or "ok"}
        tail = utf16_slice(exit.stderr.strip(JS_WHITESPACE), -STDERR_TAIL)
        error = (
            f"timed out after {timeout_ms} ms"
            if exit.timed_out
            else f"spawn failed: {exit.error}"
            if exit.error is not None
            else tail
            or f"exit code {exit.code if exit.code is not None else 'null'}"
            + (f" ({exit.signal})" if exit.signal else "")
        )
        return {
            **result,
            "verdict": headline(f"failed: {_last_line(error) or error}"),
            "error": error,
        }

    return BashHandle(proc.pid, proc.kill, asyncio.create_task(finish()))


async def run_bash_step(step: Json, opts: Json) -> Json:
    return await (await start_bash_step(step, opts)).result
